File: src/lib/utils/maintenance.ts
import { round2 } from '$lib/utils/number';
import { calculateEntryMacros } from '$lib/utils/nutrition';
import { daysBetween } from '$lib/utils/dates';
import {
	KCAL_PER_KG_FAT,
	KCAL_PER_KG_MUSCLE,
	DEFAULT_MUSCLE_RATIO
} from '../analytics/constants.generated';

export { KCAL_PER_KG_FAT, KCAL_PER_KG_MUSCLE, DEFAULT_MUSCLE_RATIO };

export type MaintenanceInput = {
	weightChangeKg: number;
	avgDailyCalories: number;
	days: number;
	muscleRatio?: number;
};

export type MaintenanceResult = {
	maintenanceCalories: number;
	dailyDeficit: number;
	totalEnergyBalance: number;
	fatMassKg: number;
	muscleMassKg: number;
	fatCalories: number;
	muscleCalories: number;
	avgDailyCalories: number;
	weightChangeKg: number;
	days: number;
	muscleRatio: number;
};

export function calculateMaintenance(input: MaintenanceInput): MaintenanceResult | null {
	const { weightChangeKg, avgDailyCalories, days, muscleRatio = DEFAULT_MUSCLE_RATIO } = input;

	if (days <= 0 || avgDailyCalories < 0) return null;

	const fatRatio = 1 - muscleRatio;
	const fatMassKg = Math.abs(weightChangeKg) * fatRatio;
	const muscleMassKg = Math.abs(weightChangeKg) * muscleRatio;

	const fatCalories = fatMassKg * KCAL_PER_KG_FAT;
	const muscleCalories = muscleMassKg * KCAL_PER_KG_MUSCLE;
	const totalEnergy = fatCalories + muscleCalories;

	const sign = weightChangeKg < 0 ? 1 : weightChangeKg > 0 ? -1 : 0;
	const totalEnergyBalance = totalEnergy * sign;
	const dailyDeficit = days > 0 ? totalEnergyBalance / days : 0;

	const maintenanceCalories = Math.round(avgDailyCalories + dailyDeficit);

	return {
		maintenanceCalories,
		dailyDeficit: Math.round(dailyDeficit),
		totalEnergyBalance: Math.round(totalEnergyBalance),
		fatMassKg: round2(fatMassKg),
		muscleMassKg: round2(muscleMassKg),
		fatCalories: Math.round(fatCalories),
		muscleCalories: Math.round(muscleCalories),
		avgDailyCalories: Math.round(avgDailyCalories),
		weightChangeKg,
		days,
		muscleRatio
	};
}

export type MaintenanceReportInput = {
	entries: Array<{
		date: string;
		calories: number | null;
		protein: number | null;
		carbs: number | null;
		fat: number | null;
		fiber: number | null;
		servings: number;
	}>;
	/** Sorted ascending by date. `entryDate` enables endpoint smoothing. */
	weights: Array<{ weightKg: number; entryDate?: string }>;
	fastingDays: Iterable<string>;
	startDate: string;
	endDate: string;
	muscleRatio?: number;
};

export type MaintenanceReport = {
	result: MaintenanceResult;
	meta: {
		weightEntries: number;
		foodEntryDays: number;
		totalDays: number;
		coverage: number;
		/** Smoothed start/end anchors (7-day means around the endpoints) when dates are known. */
		firstWeight: number;
		lastWeight: number;
		startDate: string;
		endDate: string;
	};
};

export type MaintenanceReportError = {
	error: 'insufficient_data' | 'invalid_range' | 'calculation_failed';
	message: string;
};

const ANCHOR_WINDOW_DAYS = 7;

function epochDay(date: string): number {
	return Math.floor(Date.parse(date + 'T00:00:00Z') / 86_400_000);
}

/**
 * Weight change over the interval from smoothed endpoints. A single raw
 * measurement carries up to ~2 kg of fluid noise, which over 30 days is ~500
 * kcal/day of maintenance error, so each endpoint is the mean of the weights in
 * the first / last seven days. The two anchors sit inside the interval, so
 * their difference is scaled up to the full `days` by the anchors' actual
 * separation (their mean dates). Falls back to raw endpoints when the weights
 * carry no dates or the anchors overlap.
 */
export function smoothedWeightChange(
	weights: Array<{ weightKg: number; entryDate?: string }>,
	days: number
): { firstWeight: number; lastWeight: number; weightChangeKg: number } {
	const raw = {
		firstWeight: weights[0].weightKg,
		lastWeight: weights[weights.length - 1].weightKg,
		weightChangeKg: weights[weights.length - 1].weightKg - weights[0].weightKg
	};
	if (!weights.every((w) => w.entryDate)) return raw;

	const dated = weights
		.map((w) => ({ day: epochDay(w.entryDate as string), weightKg: w.weightKg }))
		.sort((a, b) => a.day - b.day);
	const firstDay = dated[0].day;
	const lastDay = dated[dated.length - 1].day;
	const head = dated.filter((w) => w.day < firstDay + ANCHOR_WINDOW_DAYS);
	const tail = dated.filter((w) => w.day > lastDay - ANCHOR_WINDOW_DAYS);
	const meanOf = (xs: number[]) => xs.reduce((s, v) => s + v, 0) / xs.length;
	const headDay = meanOf(head.map((w) => w.day));
	const tailDay = meanOf(tail.map((w) => w.day));
	const separation = tailDay - headDay;
	if (separation <= 0) return raw;

	const firstWeight = meanOf(head.map((w) => w.weightKg));
	const lastWeight = meanOf(tail.map((w) => w.weightKg));
	return {
		firstWeight,
		lastWeight,
		weightChangeKg: ((lastWeight - firstWeight) * days) / separation
	};
}

export function buildMaintenanceReport(
	input: MaintenanceReportInput
): MaintenanceReport | MaintenanceReportError {
	const { entries, weights, fastingDays, startDate, endDate, muscleRatio } = input;

	if (weights.length < 2) {
		return {
			error: 'insufficient_data',
			message: 'At least 2 weight entries are required in the selected range'
		};
	}

	const dailyTotals: Record<string, number> = {};
	for (const entry of entries) {
		dailyTotals[entry.date] = (dailyTotals[entry.date] ?? 0) + calculateEntryMacros(entry).calories;
	}
	// Fasting days count as logged days with 0 kcal.
	for (const fastingDate of fastingDays) {
		if (!(fastingDate in dailyTotals)) dailyTotals[fastingDate] = 0;
	}

	const daysWithEntries = Object.keys(dailyTotals);
	if (daysWithEntries.length === 0) {
		return { error: 'insufficient_data', message: 'No food entries found in the selected range' };
	}

	const days = daysBetween(startDate, endDate);
	if (days <= 0) {
		return { error: 'invalid_range', message: 'End date must be after start date' };
	}

	// Mean intake is over the days that were actually logged (fasting days are
	// explicit zeros above). An unlogged day is unknown, not a zero-calorie day —
	// dividing by every calendar day understated intake, and therefore
	// maintenance, by the user's non-logging rate. The inclusive calendar count
	// only feeds the coverage figure; the weight-change *rate* keeps `days`.
	const inclusiveDays = days + 1;
	const totalCalories = Object.values(dailyTotals).reduce((sum, cal) => sum + cal, 0);
	const avgDailyCalories = totalCalories / daysWithEntries.length;
	const coverage = daysWithEntries.length / inclusiveDays;

	const { firstWeight, lastWeight, weightChangeKg } = smoothedWeightChange(weights, days);

	const result = calculateMaintenance({ weightChangeKg, avgDailyCalories, days, muscleRatio });
	if (!result) {
		return { error: 'calculation_failed', message: 'Could not calculate maintenance calories' };
	}

	return {
		result,
		meta: {
			weightEntries: weights.length,
			foodEntryDays: daysWithEntries.length,
			totalDays: inclusiveDays,
			coverage,
			firstWeight,
			lastWeight,
			startDate,
			endDate
		}
	};
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/analytics/Maintenance.kt
package com.bissbilanz.analytics

import kotlin.math.abs
import kotlin.math.floor

/** Inputs to [calculateMaintenance]; mirrors the TS `MaintenanceInput`. */
data class MaintenanceInput(
    val weightChangeKg: Double,
    val avgDailyCalories: Double,
    val days: Int,
    val muscleRatio: Double = DEFAULT_MUSCLE_RATIO,
)

/** Result of [calculateMaintenance]; mirrors the TS `MaintenanceResult` field-for-field. */
data class MaintenanceResult(
    val maintenanceCalories: Double,
    val dailyDeficit: Double,
    val totalEnergyBalance: Double,
    val fatMassKg: Double,
    val muscleMassKg: Double,
    val fatCalories: Double,
    val muscleCalories: Double,
    val avgDailyCalories: Double,
    val weightChangeKg: Double,
    val days: Int,
    val muscleRatio: Double,
)

/**
 * Estimates maintenance calories from a window's weight change and average
 * intake. Returns null for a non-positive window or negative average intake,
 * matching the server contract exactly.
 *
 * A loss (negative [MaintenanceInput.weightChangeKg]) implies intake was below
 * maintenance, so the daily deficit is added back to the average to recover
 * maintenance; a gain subtracts. All public numbers are rounded the same way
 * the TS does (JS `Math.round`, i.e. round-half-up — see [jsRound]).
 *
 * Pure port of the server's `src/lib/utils/maintenance.ts` so the mobile apps
 * can compute maintenance on-device (local/anonymous users) and online users
 * can compute it from the synced local DB without an API round-trip. The
 * golden-vector parity suite (analytics-parity/) keeps this in lockstep with
 * the TS source.
 */
fun calculateMaintenance(input: MaintenanceInput): MaintenanceResult? {
    val weightChangeKg = input.weightChangeKg
    val avgDailyCalories = input.avgDailyCalories
    val days = input.days
    val muscleRatio = input.muscleRatio

    if (days <= 0 || avgDailyCalories < 0) return null

    val fatRatio = 1 - muscleRatio
    val fatMassKg = abs(weightChangeKg) * fatRatio
    val muscleMassKg = abs(weightChangeKg) * muscleRatio

    val fatCalories = fatMassKg * KCAL_PER_KG_FAT
    val muscleCalories = muscleMassKg * KCAL_PER_KG_MUSCLE
    val totalEnergy = fatCalories + muscleCalories

    val sign =
        when {
            weightChangeKg < 0 -> 1
            weightChangeKg > 0 -> -1
            else -> 0
        }
    val totalEnergyBalance = totalEnergy * sign
    val dailyDeficit = if (days > 0) totalEnergyBalance / days else 0.0

    val maintenanceCalories = jsRound(avgDailyCalories + dailyDeficit)

    return MaintenanceResult(
        maintenanceCalories = maintenanceCalories,
        dailyDeficit = jsRound(dailyDeficit),
        totalEnergyBalance = jsRound(totalEnergyBalance),
        fatMassKg = jsRound(fatMassKg * 100) / 100,
        muscleMassKg = jsRound(muscleMassKg * 100) / 100,
        fatCalories = jsRound(fatCalories),
        muscleCalories = jsRound(muscleCalories),
        avgDailyCalories = jsRound(avgDailyCalories),
        weightChangeKg = weightChangeKg,
        days = days,
        muscleRatio = muscleRatio,
    )
}

/**
 * JavaScript `Math.round` semantics: round half toward +infinity (`Math.round(2.5)
 * == 3`, `Math.round(-2.5) == -2`). Kotlin's `kotlin.math.round` rounds half to
 * even, so it cannot be used here without breaking byte-for-byte parity with the
 * TS server.
 */
internal fun jsRound(value: Double): Double = floor(value + 0.5)

/** A weight measurement with an optional date, for [smoothedWeightChange]. */
data class DatedWeight(
    val weightKg: Double,
    val entryDate: String? = null,
)

data class WeightChange(
    val firstWeight: Double,
    val lastWeight: Double,
    val weightChangeKg: Double,
)

private const val ANCHOR_WINDOW_DAYS = 7

/**
 * Weight change over the interval from smoothed endpoints. A single raw
 * measurement carries up to ~2 kg of fluid noise, so each endpoint is the mean
 * of the weights in the first / last seven days; the anchors sit inside the
 * interval, so their difference is scaled up to the full [days] by the anchors'
 * actual separation. Falls back to raw endpoints when the weights carry no
 * dates or the anchors overlap. Mirrors the TS `smoothedWeightChange`.
 */
fun smoothedWeightChange(
    weights: List<DatedWeight>,
    days: Int,
): WeightChange {
    val raw = WeightChange(weights.first().weightKg, weights.last().weightKg, weights.last().weightKg - weights.first().weightKg)
    if (weights.any { it.entryDate == null }) return raw

    val dated =
        weights
            .map {
                Pair(
                    kotlinx.datetime.LocalDate
                        .parse(it.entryDate!!)
                        .toEpochDays(),
                    it.weightKg,
                )
            }.sortedBy { it.first }
    val firstDay = dated.first().first
    val lastDay = dated.last().first
    val head = dated.filter { it.first < firstDay + ANCHOR_WINDOW_DAYS }
    val tail = dated.filter { it.first > lastDay - ANCHOR_WINDOW_DAYS }
    val headDay = head.sumOf { it.first.toDouble() } / head.size
    val tailDay = tail.sumOf { it.first.toDouble() } / tail.size
    val separation = tailDay - headDay
    if (separation <= 0) return raw

    val firstWeight = head.sumOf { it.second } / head.size
    val lastWeight = tail.sumOf { it.second } / tail.size
    return WeightChange(firstWeight, lastWeight, ((lastWeight - firstWeight) * days) / separation)
}
